Add widerface dataset

jgbradley1 opened the pull request:
This PR addresses issue #1627 and adds a dataset that is more representative (and harder) of generic face detection.

Main changes in the PR include
* adds WIDERFace dataset as an option
* <strike>improves the method to check and download files from google drive when download quota has been exceeded. See the Google drive API [docs](https://developers.google.com/drive/api/v3/handle-errors#:~:text=An%20error%20403%20occurs%20when,The%20daily%20limit%20was%20exceeded.) for more details, A quota limit exceeded will return a 403 response. Checking for the correct response code before performing a string search saves a significant amount of time for the most general use case (i.e. when the quota limit has not been exceeded).</strike>

cc: @pmeier @vfdev-5

pmeier reviewed (commented):
Hi @jgbradley1 and thanks a lot for this PR! Before you commit any more time into this, please note that we aren't sure yet if we want to add the WIDERFace dataset. We are currently in the process of adding [contributing guidelines](https://github.com/pytorch/vision/pull/2663/) including new datasets. Right now our only metric is the number of citations:

- CelebA: [86](https://scholar.google.com/scholar?cites=286831505508630744)
- WIDERFace: [682](https://scholar.google.com/scholar?cites=8866787776930184069)

Given that WIDERFace seems to be used more than CelebA, I think it is reasonable to include it. That being said, I have no final say here. I suggest you hold further development until @fmassa gives green light for this. If he does I'm happy to review.

jgbradley1 commented:
@fmassa Have you had time to consider this new dataset? The contributing guidelines do not include guidance yet regarding dataset requests.

fmassa commented on torchvision/datasets/widerface.py:
```diff
@@ -0,0 +1,231 @@
+        filename = "wider_face_train_bbx_gt.txt" if self.split == "train" else "wider_face_val_bbx_gt.txt"
+        filepath = os.path.join(self.root, "wider_face_split", filename)
+
+        f = open(filepath, "r")
```
nit: can you use the context manager to close the file after it being opened?

fmassa reviewed (commented):
Hi @jgbradley1 

Sorry for the delay.

We would love to have WIDERFACE added to torchvision, thanks for the PR!

One meta-comment I would have on the structure of the PR:
- we would prefer to avoid having `target_type` to define what to return from the dataset. I think it would be preferable instead to always return the `"raw"` mode as a dict. I know we have `target_type` in CelebA already, but I would prefer if the output type of the datasets was always the same.
- having tests for the dataset (which creates a dummy data instance and check that everything works as expected) would be great

I'll let @pmeier have a closer look at the implementation if he has other comments.

pmeier commented:
@jgbradley1 I'll review later today.

pmeier commented on torchvision/datasets/widerface.py:
```diff
@@ -0,0 +1,231 @@
+            downloaded again.
+    """
+
+    base_folder = "widerface"
+    file_list = [
+        # File ID                        MD5 Hash                            Filename
+        ("0B6eKvaijfFUDQUUwd21EckhUbWs", "3fedf70df600953d25982bcd13d91ba2", "WIDER_train.zip"),
+        ("0B6eKvaijfFUDd3dIRmpvSk8tLUk", "dfa7d7e790efa35df3788964cf0bbaea", "WIDER_val.zip"),
+        ("0B6eKvaijfFUDbW4tdGpaYjgzZkU", "e5d8f4248ed24c334bbd12f49c29dd40", "WIDER_test.zip")
+    ]
+    annotations_file = ("http://mmlab.ie.cuhk.edu.hk/projects/WIDERFace/support/bbx_annotation/wider_face_split.zip",
```
To indicate that these are constants, could you capitalize them?

jgbradley1 replied:
No  problem. I was following conventions found in other dataset classes, but capitalization makes sense here.

pmeier replied:
We'll happily accept PRs changing that :wink:

pmeier commented on torchvision/datasets/widerface.py:
```diff
@@ -0,0 +1,231 @@
+            filepath = os.path.join(self.root, filename)
+            if not check_integrity(filepath, md5):
+                return False
+            extract_archive(filepath)
```
Unsure about this. Why would `check_integrity` extract the archive? Shouldn't this at least be in download?

jgbradley1 replied:
This covers the use case where someone has manually downloaded the zipped files to a filesystem but forgot to or has not yet unzipped the files. In this scenario, I do not expect the user to call `WIDERFace` with `download=True`...and we must still check and decompress the files if their respective folders do not exist.

pmeier replied:
> In this scenario, I do not expect the user to call `WIDERFace` with `download=True`

That depends on the interpretation of `download=True`. For me (and this is how we handled this in other datasets) `download=True` means "please download **and** prepare the dataset so that I can use it". We could split this into `download` and `prepare` to be even more explicit. 

Still, IMO `check_integrity` is not the right place to put this. From the name I only expect it to check if everything is alright and not process the data in any way.

jgbradley1 replied:
@pmeier I've moved the dataset preparation logic so `download` downloads __and__ extracts the dataset files. `check_integrity(...)` behaves the way you describe now.

Unit tests have been added as well. I had to disable the tests on windows, otherwise error messages were raised regarding file permission errors when removing the dummy dataset instance. Whatever the issue is, it's outside the scope of the test based on what I can tell...I suspect this is the same reason why the cityscapes dataset tests are disabled on windows.

pmeier commented on torchvision/datasets/widerface.py:
```diff
@@ -0,0 +1,231 @@
+class WIDERFace(VisionDataset):
+    """`WIDERFace <http://shuoyang1213.me/WIDERFACE/>`_ Dataset.
+
+    Citation:
+    @inproceedings{yang2016wider,
+        author    = "Yang, Shuo and Luo, Ping and Loy, Chen Change and Tang, Xiaoou",
+        booktitle = "IEEE Conference on Computer Vision and Pattern Recognition (CVPR)",
+        title     = "WIDER FACE: A Face Detection Benchmark",
+        year      = "2016"}
```
This would set precedence, since we do not have a citation in any other dataset. To be clear, I like the idea ("credit were credit is due"), but something like this entails a lot. @fmassa?

jgbradley1 replied:
I included this for two primary reasons
* to direct users to the original paper for questions about how to interpret the dataset
* there are several version of widerface annotations floating around on the internet with a different folder structure. I want to make it clear that the annotations are the original annotations released with the dataset.

If this has implications like altering the license in someway, I'm happy to leave it off. I just figured the docstring would be more helpful by directing people to the original source.

pmeier reviewed (changes requested):
I have some minor suggestions and questions. Other than that, I second @fmassa s two points above. If you need help designing or writing the tests, let me know.

jgbradley1 commented on torchvision/datasets/widerface.py:
```diff
@@ -0,0 +1,231 @@
+                                        target_transform=target_transform)
+        # check arguments
+        if split not in ("train", "val", "test"):
+            raise ValueError("split \"{}\" is not recognized.".format(split))
```
You are correct. I wasn't sure how to use it before but it makes sense now. Switched over to using `verify_str_arg`

jgbradley1 commented on torchvision/datasets/widerface.py:
```diff
@@ -0,0 +1,231 @@
+        filepath = os.path.join(self.root, "wider_face_split", filename)
+
+        f = open(filepath, "r")
+        lines = f.readlines()
```
Switched over to using a context manager for file operations.

jgbradley1 commented on torchvision/datasets/utils.py:
```diff
@@ -124,7 +124,7 @@ def _quota_exceeded(response: "requests.models.Response") -> bool:  # type: igno
 
 def download_file_from_google_drive(file_id: str, root: str, filename: Optional[str] = None, md5: Optional[str] = None):
-    """Download a Google Drive file from  and place it in root.
+    """Download a Google Drive file and place it in root.
```
I'll submit a follow-up PR for the typo fix. Change has been reverted.

jgbradley1 commented on torchvision/datasets/widerface.py:
```diff
@@ -0,0 +1,231 @@
+                target = None
+                break
+            else:
+                raise ValueError("Target type \"{}\" is not recognized.".format(t))
```
Yes. I thought explicitly listing all if-else case conditions would be easier to read than a catch-all else statement at the end (even though the else condition would not be reached in theory). Code has been refactored with a simple line comment to explain the else condition.

jgbradley1 commented:
> I think it would be preferable instead to always return the "raw" mode as a dict. I know we have target_type in CelebA already, but I would prefer if the output type of the datasets was always the same.

@fmassa I need some clarification. Are you suggesting to remove the `target_type` argument entirely and always return a dictionary? For example, the raw annotation for one image with one face would be:
```
{
  "raw": tensor([[983, 296, 7, 9, 2, 0, 0, 1, 1, 0]]), # torch.Size([1, 10])
  "bbox": tensor([[983, 296, 7, 9]]),                  # torch.Size([1, 4])
  "attributes": tensor([[2, 0, 0, 1, 1, 0]])           # torch.Size([1, 6])
}
```
Should the `raw` key-value pair be left as part of the dict? The tensors will change shape based on the number of faces in an image but hopefully you get the idea from the example above.

If `target_type` is removed, this may change how `target_transform` is handled. I used `CelebA` as a template for applying `target_transform`. Is there an better Dataset class I should reference as the standard?

jgbradley1 commented:
@fmassa and @pmeier I'm just picking this PR back up. It got lost during the holidays I think.

I've removed the dataset citation for now and removed the `target_type` argument as requested earlier. Also updated the output type to always be a `dict`.

Do you have any other suggestions or is this about ready to merge?

pmeier reviewed (commented):
If I understand your implementation correctly, the target is always returned as `None` for the test split. If that is the case, I would like to hear @fmassa s thoughts about this, since it would create precedence. As of now, all our datasets always return a valid target.

Other than that two minor nitpicks.

jgbradley1 commented on torchvision/datasets/widerface.py:
```diff
@@ -0,0 +1,183 @@
+
+        # process dataset
+        # dataset will be stored as a list of dict objects (img_info)
+        self.img_info: Any = []
```
I've removed those two comment lines entirely as the function names are pretty self-documenting.

pmeier replied:
My bad for not being precise enough: I meant the type annotation `Any`. At least it should be `List[Any]` but could probably be even more precise.

jgbradley1 replied:
Would this be precise enough?

```suggestion
        self.img_info: List[Dict[str, Any]] = []
```

pmeier replied:
Depends :wink: What type do the values of the dicts have? As a rule of thumb one should only use `Any` if it is indeed not important or simply not expressible.

In this case here IMO this would be the correct annotation:

```suggestion
        self.img_info: List[Dict[str, Union[str, Dict[str, torch.Tensor]]]] = []
```

jgbradley1 replied:
You are correct. I've made the change you suggested. Resolving for conversation for now but feel free to open back up if you have any other feedback.

jgbradley1 commented:
> If I understand your implementation correctly, the target is always returned as `None` for the test split. If that is the case, I would like to hear @fmassa s thoughts about this, since it would create precedence. As of now, all our datasets always return a valid target.

Ground truth annotations for the WIDERFace test split are not released publicly (as noted in the [description section](http://shuoyang1213.me/WIDERFACE/#Description) of the main website). In this case, the only options we can provide here is either:
1) drop support for the test split entirely or
2) support reading the test split (to improve ease of evaluation) and return an empty target

Actual evaluation on the test split requires a submission to the dataset author. I prefer option 2 as it would encourage users to go for SOTA on the dataset.

fmassa commented:
Hi @jgbradley1 

Sorry for the delay.

I think I would be ok letting the dataset return a `target=None` for the test split. This way, the return types are always consistent.

If we were to return either `image` or `image, target`, then I would be a bit more reticent to do it, but returning `image, None` I think it's fine.

jgbradley1 commented:
Looks like there is a problem with a couple of the conda CI builds (unrelated to my changes as far I can tell). I don't think it's anything I can resolve from my end.

pmeier reviewed (approved):
Thanks a lot @jgbradley1!
